### PR TITLE
feat(pi05): add native FSDP support via JointDecoderLayerPair

### LINKED ISSUE #3562: Does LeRobot support multi-GPU training with FSDP or DeepSpeed ZeRO-2?
### Ticket Type

❓ Technical Question

### Environment & System Info

```Shell
- LeRobot version: 0.5.2
- Platform: Linux-6.8.0-79-generic-x86_64-with-glibc2.43
- Python version: 3.12.13
- Huggingface Hub version: 1.14.0
- Transformers version: 5.5.4
- Datasets version: 4.8.5
- Numpy version: 2.2.6
- FFmpeg version: 8.1.1
- PyTorch version: 2.10.0+cu128
- Is PyTorch built with CUDA support?: True
- Cuda version: 12.8
- GPU model: NVIDIA A100-PCIE-40GB
- Using GPU in script?: <fill in>
- lerobot scripts: ['lerobot-calibrate', 'lerobot-dataset-viz', 'lerobot-edit-dataset', 'lerobot-eval', 'lerobot-find-cameras', 'lerobot-find-joint-limits', 'lerobot-find-port', 'lerobot-imgtransform-viz', 'lerobot-info', 'lerobot-record', 'lerobot-replay', 'lerobot-rollout', 'lerobot-setup-can', 'lerobot-setup-motors', 'lerobot-teleoperate', 'lerobot-train', 'lerobot-train-tokenizer']
```

### Description

Hi LeRobot team, thanks for the great project!

I would like to ask whether LeRobot currently supports multi-GPU training with either:

- PyTorch FSDP
- DeepSpeed ZeRO Stage 2

I am trying to scale training to multiple GPUs, but so far I have not been able to get FSDP or DeepSpeed ZeRO-2 working successfully with the current training pipeline.

Could you clarify whether these distributed training strategies are currently supported or expected to work with LeRobot?

If they are supported, it would be very helpful to have a minimal working example or recommended launch command/configuration.

### Context & Reproduction

_No response_

### Relevant logs or stack trace

```Shell

```

### Checklist

- [x] I have searched existing tickets to ensure this isn't a duplicate.
- [x] I am using the latest version of the `main` branch.
- [x] I have verified this is not an environment-specific problem.

### Additional Info / Workarounds

_No response_

### PR DESCRIPTION
**Pull request by nihal111**
## Summary / Motivation

Pi0.5's forward pass calls `compute_layer_complete()` as a free function that directly accesses weight matrices from paired PaliGemma/Expert layers without calling `module.forward()`. FSDP relies on `forward()` hooks to trigger all-gather of sharded parameters, so weights remain sharded when accessed, producing dimension mismatches. There is currently no working PyTorch FSDP implementation for Pi0.5 in LeRobot, OpenPI, or the community (see #3562).

This PR adds `JointDecoderLayerPair(nn.Module)` which wraps each paired decoder layer and provides the `forward()` boundary FSDP needs. The 18 pairs (~134M params each) become natural FSDP sharding units via `transformer_auto_wrap_policy`. With FSDP FULL_SHARD, per-GPU memory drops from ~49 GB to much lesser (depending on extent of sharding), enabling full fine-tuning on more commonly available 40GB GPU clusters.

## Related issues

- Closes #3562
- Related: #2216 (GPU memory for pi05), #2154 (Multi-GPU DDP)

## What changed

- **`JointDecoderLayerPair(nn.Module)`**: Wraps each paired (PaliGemma layer_i, Expert layer_i) and delegates `forward()` to `compute_layer_complete()`. FSDP wraps each pair as one sharding unit.
- **`_JointLayerView(nn.Module)`**: Non-owning proxy that replaces the emptied original layer lists. The HuggingFace inference path (prefix-only/suffix-only with KV cache) still iterates `self.layers` unchanged.
- **`_build_joint_layers()`** on `PaliGemmaWithExpertModel`: Called in `__init__`, moves layers into `joint_layers` ModuleList, replaces originals with views.
- **Forward rewrite**: The training path (`else` branch in `PaliGemmaWithExpertModel.forward`) now iterates `self.joint_layers` calling `joint_layer(...)` instead of zipping separate lists and calling the free function directly.
- **State dict key remapping** in `_fix_pytorch_state_dict_keys`: Old checkpoint keys (`paligemma.model.language_model.layers.N...` / `gemma_expert.model.layers.N...`) remap to `joint_layers.N.paligemma_layer...` / `joint_layers.N.expert_layer...` for full backward compatibility.
- **No config changes required**: FSDP users just set `fsdp_transformer_layer_cls_to_wrap: "JointDecoderLayerPair"` in their accelerate config.

This is NOT a breaking change for DDP or single-GPU users. The model produces identical outputs; only the module tree structure changes.

## How was this tested (or how to run locally)

Tests added: `tests/policies/pi0_pi05/test_pi05_fsdp.py`

```bash
pytest -q tests/policies/pi0_pi05/test_pi05_fsdp.py
```

Covers:
- `TestJointLayerStructure`: 18 pairs created, view proxying, no param duplication, state_dict keys correct
- `TestFSDPWrapPolicy`: auto-wrap identifies 18 units, params per pair in expected range
- `TestForwardPass`: training forward (loss shape), inference forward (sample_actions with KV cache)
- `TestCheckpointCompatibility`: old-format keys remap and strict-load with 0 missing/unexpected

Manual validation (CPU, no FSDP runtime):
- Model instantiation: 4.14B params, no duplication
- Training forward pass: correct loss shape
- Inference with KV cache (sample_actions, num_steps=2): correct action shape
- FSDP wrap policy: 18 `JointDecoderLayerPair` modules found at expected paths
- Checkpoint round-trip: old-format state_dict -> key remap -> strict load succeeds
- `ruff check` passes clean

Additionally validated on a 10-node x 8 A100-40GB cluster (80 GPUs) using FSDP FULL_SHARD via accelerate, training for 14+ hours / 2500 steps with loss converging from 0.377 to 0.037.

## Checklist (required before merge)

- [x] Linting/formatting run (`ruff check` passes)
- [ ] All tests pass locally (`pytest`) - pi05 tests require CUDA + HF token; structural tests pass on CPU
- [ ] Documentation updated
- [ ] CI is green
- [ ] Community Review: I have reviewed another contributor's open PR and linked it here: #

## Reviewer notes

- The `_JointLayerView` class is the most subtle piece: it must behave like an `nn.ModuleList` for iteration/indexing (HuggingFace's `PiGemmaModel.forward` iterates `self.layers[: self.config.num_hidden_layers]`) while NOT registering child modules (to avoid double-counting params). Worth careful review.
- The `rotary_emb` reference is stored via `object.__setattr__` to avoid duplicate module registration in the tree (the instance is already registered under `paligemma.model.language_model.rotary_emb`).
- Accelerate FSDP config example for users:
  ```yaml
  fsdp_config:
    fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
    fsdp_transformer_layer_cls_to_wrap: "JointDecoderLayerPair"
    fsdp_sharding_strategy: FULL_SHARD
    fsdp_state_dict_type: SHARDED_STATE_DICT
    fsdp_sync_module_states: true
    fsdp_use_orig_params: true
  ```

### CODE REVIEW

**Comment by nihal111** [comment]
@claude review
